src/lib/mina/test/test-contract.ts

```diff
@@ -0,0 +1,319 @@
+    if (maybe !== undefined) {
+      if (!Array.isArray(maybe)) maybe = [maybe];
+      for (let instruction of maybe) await runInstruction(spec, instruction);
+    }
+  } else if (instruction.type === 'transaction') {
+    console.time(instruction.label);
+    let feepayer = instruction.sender ?? sender;
+    let signers = [feepayer.key, ...(instruction.signers ?? [])];
+    let tx = await Mina.transaction(feepayer, instruction.callback);
+    await assertionWithTrace(instruction.trace, async () => {
+      // console.log(instruction.label, tx.toPretty());
+      await tx.sign(signers).prove();
+      await tx.send();
+    });
+    console.timeEnd(instruction.label);
+  } else if (instruction.type === 'deploy') {
+    let { options, trace } = instruction;
+    let account = options?.account ?? contractAccount;
+    let contract = options?.contract ?? Contract;
+    let instance =
+      contract instanceof SmartContract ? contract : new contract(account);
+
+    await runInstruction(spec, {
+      type: 'transaction',
+      label: 'deploy',
+      callback: () => instance.deploy(),
+      trace,
+      sender,
+      signers: [account.key],
+    });
+  } else if (instruction.type === 'expect-state') {
+    let { state, expected, trace, label } = instruction;
+    if ('_type' in state) {
+      let type = state._type;
+      await assertionWithTrace(trace, async () => {
+        let actual = Option(type).toValue(await state.get());
+        assert.deepStrictEqual(actual, expected, label);
+      });
+    } else if ('_valueType' in state) {
+      let [key, value] = expected;
+      let type = state._valueType;
+      await assertionWithTrace(trace, async () => {
+        let actual = Option(type).toValue(await state.get(key));
+        assert.deepStrictEqual(actual, value, label);
+      });
+    }
+  } else if (instruction.type === 'expect-balance') {
+    let { address, expected, label, trace } = instruction;
+    await assertionWithTrace(trace, () => {
+      let actual = Mina.getBalance(address).toBigInt();
+      assert.deepStrictEqual(actual, expected, label);
+    });
+  } else {
+    throw new Error('Unknown test instruction type');
+  }
+}
+
+// types and helper structures
+
+type MaybePromise<T> = T | Promise<T>;
+
+type BaseInstruction = { type: string; trace?: string; label?: string };
+
+type TestInstruction =
+  | ((...args: any) => MaybePromise<TestInstruction | TestInstruction[] | void>)
+  | (BaseInstruction &
+      (
+        | {
+            type: 'transaction';
+            label: string;
+            callback: () => Promise<void>;
+            sender?: Mina.TestPublicKey;
+            signers?: PrivateKey[];
+          }
+        | {
+            type: 'deploy';
+            options?: {
+              contract?: typeof SmartContract | SmartContract;
+              account?: Mina.TestPublicKey;
+            };
+          }
+        | { type: 'expect-state'; state: State; expected: Expected<State> }
```

Comment:
What do you think about adding more `expect` cases here? A few ideas I can think of a few that could be helpful:

1. Expect token balance
- `{ type: 'expect-token-balance'; address: PublicKey; tokenId: Field; expected: bigint }`
2. Expect event
- `{ type: 'expect-event'; eventName: string; data?: any }`
3. Expect action
- `{ type: 'expect-action'; data?: any }`
4. Expect error
- `{ type: 'expect-error'; callback: () => Promise<void>; errorType?: typeof Error }`

Feel free to ignore if they don't seem interesting/worthwhile. Could always be added later as well

Reply:
great point

expect error especially, that's completely missing from the current framework

Reply:
but let's defer making the test framework production-ready to a different PR